tools/ci-trigger/pr.go

```diff
@@ -0,0 +1,334 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/go-git/go-git/v5"
+	"github.com/golang/glog"
+	"github.com/google/go-github/v50/github"
+	"google.golang.org/api/cloudbuild/v1"
+	"google.golang.org/protobuf/encoding/prototext"
+
+	mpb "github.com/openconfig/featureprofiles/proto/metadata_go_proto"
+	opb "github.com/openconfig/ondatra/proto"
+)
+
+type pullRequest struct {
+	ID       int
+	CloneURL string
+	HeadSHA  string
+	BaseSHA  string
+	Virtual  []device
+	Physical []device
+
+	repo      *git.Repository
+	localFS   fs.FS
+	localPath string
+}
+
+type device struct {
+	Type             deviceType
+	CloudBuildID     string
+	CloudBuildLogURL string
+	Tests            []functionalTest
+}
+
+type deviceType struct {
+	Vendor        opb.Device_Vendor
+	HardwareModel string
+}
+
+type functionalTest struct {
+	Name        string
+	Description string
+	Path        string
+	DocURL      string
+	TestURL     string
+	Status      string
+	BadgePath   string
+	BadgeURL    string
+}
+
+func (d *deviceType) String() string {
+	return d.Vendor.String() + " " + d.HardwareModel
+}
+
+// CreateBuild creates a GCB build for each of the deviceTypes.
+func (p *pullRequest) CreateBuild(ctx context.Context, buildClient *cloudbuild.Service, storClient *storage.Client, devices []deviceType) error {
+	for _, d := range devices {
+		for i, virtualDevice := range p.Virtual {
+			if virtualDevice.Type == d {
+				skip := false
+				for _, v := range virtualDevice.Tests {
+					if v.Status != "pending authorization" {
+						skip = true
+					}
+				}
+				if skip {
+					continue
+				}
+				cb := newCloudBuild(p.localFS, buildClient, storClient, virtualDevice)
+				jobID, logURL, err := cb.SubmitBuild(ctx)
+				if err != nil {
+					return fmt.Errorf("error creating CloudBuild job for PR%d: %w", p.ID, err)
+				}
+				p.Virtual[i].CloudBuildID = jobID
+				p.Virtual[i].CloudBuildLogURL = logURL
+				for j := range virtualDevice.Tests {
+					p.Virtual[i].Tests[j].Status = "setup"
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// IdentifyModifiedTests gathers all of the tests that have been modified in the pull request.
+func (p *pullRequest) IdentifyModifiedTests() error {
+	var err error
+
+	if p.repo == nil {
+		p.repo, err = setupGitClone(p.localPath, p.CloneURL, p.HeadSHA)
+		if err != nil {
+			return err
+		}
+	}
+
+	ft, err := functionalTestPaths(p.localFS)
+	if err != nil {
+		return err
+	}
+
+	mf, err := modifiedFiles(p.repo, p.HeadSHA, p.BaseSHA)
+	if err != nil {
+		return err
+	}
+	modifiedTests := modifiedFunctionalTests(ft, mf)
+
+	err = p.populateTestDetail(modifiedTests)
+	if err != nil {
+		return nil
```

Comment:
return err?

could simplify this to just:
return p.populateTestDetail(modifiedTests)